libimage/runtime.go

```diff
@@ -574,23 +574,30 @@ func (r *Runtime) ListImages(ctx context.Context, names []string, options *ListI
 		options = &ListImagesOptions{}
 	}
 
-	var images []*Image
-	if len(names) > 0 {
-		for _, name := range names {
-			image, _, err := r.LookupImage(name, nil)
-			if err != nil {
-				return nil, err
-			}
-			images = append(images, image)
-		}
-	} else {
-		storageImages, err := r.store.Images()
+	lookupImagesMap := make(map[string]struct{}, 0)
+	for _, name := range names {
+		image, _, err := r.LookupImage(name, nil)
 		if err != nil {
 			return nil, err
 		}
-		for i := range storageImages {
-			images = append(images, r.storageToImage(&storageImages[i], nil))
+		lookupImagesMap[image.ID()] = struct{}{}
+	}
+
+	storageListResult, err := r.store.MultiList(
+		storage.MultiListOptions{Images: true, Layers: true},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	images := make([]*Image, 0, len(storageListResult.Images))
+	for i := range storageListResult.Images {
+		image := r.storageToImage(&storageListResult.Images[i], nil)
+
+		if _, found := lookupImagesMap[image.ID()]; !found && len(names) > 0 {
+			continue
```

Comment:
(Style nit: `continue` is sometimes the right tool, but often using a structured `if` is cleaner, for the same reason we generally shy away from `goto`. Admittedly others feel differently and prefer `continue`-heavy code.)